controllers/topology/cluster_controller_test.go

```diff
@@ -0,0 +1,377 @@
+		WithInfrastructureClusterTemplate(infrastructureClusterTemplate).
+		WithControlPlaneTemplate(controlPlaneTemplate).
+		WithControlPlaneInfrastructureMachineTemplate(infrastructureMachineTemplate).
+		WithWorkerMachineDeploymentClasses([]clusterv1.MachineDeploymentClass{*machineDeploymentClass, *machineDeploymentClass2}).
+		Build()
+
+	// 3) A Cluster object including a Cluster Topology Object and the MachineDeploymentTopology objects used in the
+	// Cluster Topology.
+	machineDeploymentTopology := builder.MachineDeploymentTopology("mdm").
+		WithClass(workerClassName).
+		WithReplicas(3).
+		Build()
+	machineDeploymentTopology2 := builder.MachineDeploymentTopology("mdm2").
+		WithClass(workerClassName).
+		WithReplicas(1).
+		Build()
+	cluster := builder.Cluster(ns.Name, clusterName).
+		WithTopology(
+			builder.ClusterTopology().
+				WithClass(clusterClass.Name).
+				WithMachineDeployment(machineDeploymentTopology).
+				WithMachineDeployment(machineDeploymentTopology2).
+				WithVersion("1.22.2").
+				WithControlPlaneReplicas(3).
+				Build()).
+		Build()
+
+	// Create a set of initObjects from the objects above to add to the API server when the test environment starts.
+	initObjs := []client.Object{
+		clusterClass,
+		cluster,
+		infrastructureClusterTemplate,
+		infrastructureMachineTemplate,
+		bootstrapTemplate,
+		controlPlaneTemplate,
+	}
+	defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterTopology, true)()
+
+	for _, obj := range initObjs {
+		err := env.Create(ctx, obj)
+		g.Expect(err).ToNot(HaveOccurred())
+	}
+	defer func() {
+		for _, obj := range initObjs {
+			err := env.Delete(ctx, obj)
+			g.Expect(err).ToNot(HaveOccurred())
+		}
+	}()
+	actualCluster := &clusterv1.Cluster{}
+	// Get Cluster object and check that relevant fields are set as expected.
+	g.Eventually(func() bool {
+		return assertClusterReconcile(ns.Name, clusterName, actualCluster)
+	}, timeout).Should(BeTrue())
+
+	// Check if InfrastructureCluster has been created and has the correct labels and annotations.
+	g.Eventually(func() bool {
+		return assertInfrastructureClusterReconcile(actualCluster)
+	}, timeout).Should(BeTrue())
+
+	// Check if ControlPlane has been created and has the correct version, replicas, labels and annotations.
+	g.Eventually(func() bool {
+		return assertControlPlaneReconcile(actualCluster, clusterClass)
+	}, timeout).Should(BeTrue())
+	g.Eventually(func() bool {
+		return assertMachineDeploymentsReconcile(actualCluster)
+	}, timeout).Should(BeTrue())
+}
+
+// assertClusterReconcile checks if the Cluster object:
+// 1) Has its InfrastructureReference and ControlPlane reference added correctly.
+// 2) InfrastructureReference and ControlPlaneRef have the expected Group, Version and Kind.
+func assertClusterReconcile(namespace, clusterName string, cluster *clusterv1.Cluster) bool {
+	key := client.ObjectKey{Name: clusterName, Namespace: namespace}
+	if err := env.Get(ctx, key, cluster); err != nil {
```

Comment:
Can we maybe retrieve the cluster outside of this func? Those kind of output input parameters are an anti-pattern imho as it makes code harder to read because it's surprising to have input parameters which are used as return parameters

Reply:
Agreed - will update.